rescript/tests/test_get_unite.py

```diff
@@ -0,0 +1,88 @@
+from rescript.get_unite import (
+    UNITE_DOIS,
+    _unite_get_url,
+    _unite_get_tgz,
+    _unite_get_artifacts,
+    get_unite_data,
+)
+
+from urllib.request import urlopen
+from unittest.mock import patch
+
+
+class TestGetUNITE(TestPluginBase):
+    package = "rescript.tests"
+
+    def setUp(self):
+        super().setUp()
+        self.unitefile = pkg_resources.resource_filename(
+            "rescript.tests", "data/unitefile.tgz"
+        )
+        self.unitefile_no_dev = pkg_resources.resource_filename(
+            "rescript.tests", "data/unitefile_no_dev.tgz"
+        )
+
+    # Requires internet access
+    def test_unite_get_url(self):
+        # for all combinations...
+        for v in UNITE_DOIS.keys():
+            for tg in UNITE_DOIS[v].keys():
+                for s in UNITE_DOIS[v][tg].keys():
+                    # ... try to get the URL
+                    url = _unite_get_url(v, tg, s)
+                    urlopen(url)
+        self.assertTrue(True)
+
+    # Requires internet access
+    def test_unite_get_tgz(self):
+        # Download a single, small, unrelated file for testing
+        url = "https://files.plutof.ut.ee/doi/C9/F6/C9F687C997F72F674AA539CB80BF5D5BF6D1F402A2ACF840B20322850D3DFBA4.zip"  # noqa E501
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            # Test working URL
+            _unite_get_tgz(url, tmpdirname)
+            # Test bad url
+            with self.assertRaises(ValueError):
+                _unite_get_tgz("https://files.plutof.ut.ee/nope", tmpdirname)
+
+    def test_unite_get_artifacts(self):
+        # Test on small data/unitefile.tgz with two items inside
+        res_one, res_two = _unite_get_artifacts(
+            self.unitefile, cluster_id="97"
+        )
+        # Column names and one feature from TaxonomyFormat
+        self.assertEqual(
+            res_one["Taxon"]["SH1140752.08FU_UDB013072_reps"],
+            "k__Fungi;p__Basidiomycota;c__Agaricomycetes;o__Thelephorales;"
+            "f__Thelephoraceae;g__Tomentella;s__unidentified",
+        )
+        self.assertEqual(
+            str(type(res_two)),
+            "<class 'q2_types.feature_data._transformer.DNAIterator'>",
+        )
+        # test no _dev files found
+        with self.assertRaises(ValueError):
+            _unite_get_artifacts(self.unitefile_no_dev, cluster_id="97")
+        # test missing files or misspelled cluster_id
+        with self.assertRaises(ValueError):
+            _unite_get_artifacts(self.unitefile, "nothing")
+
+    # This tests the function with toy data.
+    # All relevant internals are tested elsewhere in this test class.
+    # Downloading is mock`ed with patch.
+    def test_get_unite_data(self):
+        with patch(
+            "rescript.get_unite._unite_get_tgz", return_value=self.unitefile
+        ):
+            get_unite_data(version="8.3", taxon_group="fungi", cluster_id="97")
+            self.assertTrue(True)
```

Comment:
I would not use `self.assertTrue(True)` here. You should be able to implement something similar to [these `self.assertEqual` lines](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/tests/test_get_gtdb.py#L110).

Reply:
The function `get_unite_data` directly returns `return _unite_get_artifacts(tar_file_path, cluster_id)`

I test that above
```python
        res_one, res_two = _unite_get_artifacts(
            self.unitefile, cluster_id="97"
        )
        # Column names and one feature from TaxonomyFormat
        self.assertEqual(
            res_one["Taxon"]["SH1140752.08FU_UDB013072_reps"],
            "k__Fungi;p__Basidiomycota;c__Agaricomycetes;o__Thelephorales;"
            "f__Thelephoraceae;g__Tomentella;s__unidentified",
        )
        self.assertEqual(
            str(type(res_two)),
            "<class 'q2_types.feature_data._transformer.DNAIterator'>",
        )
```

What _should_ be testing in these two function, so they don't overlap too, too much?

Reply:
Also, their API is down for everything right now. This makes the HTML 500 error I could not replicate before, which implies I should add back in the HTML error check, or just wait until it comes back online

Reply:
get_gtdb() is a pipeline, while get_unite_data() is a function

The test would look like this

```python
        res = get_unite_data(version="8.3", taxon_group="fungi", cluster_id="97")
        self.assertEqual(len(res), 2)
        self.assertEqual(str(type(res[0])), "<class 'pandas.core.frame.DataFrame'>")
        self.assertEqual(str(type(res[1])), "<class 'q2_types.feature_data._transformer.DNAIterator'>"),
# or
        self.assertTrue(isinstance(res[0], (pandas.core.frame.DataFrame)))
        self.assertTrue(isinstance(res[1], (q2_types.feature_data._transformer.DNAIterator)))
```

Want me to add those?

Reply:
Right! I think either will do. Just check that we are getting the expected data format, for the step we're going for. I like the later ones, *i.e.* `...(pandas.core.frame.DataFrame)...`

Reply:
(Here's what I was missing. Right now, the main function get_unite_data() simply returns _unite_get_artifacts(), but if this changes, we want to make sure the output types are consistent, necessitating this ~~duplicate~~ repeated test.)

New function, new test. Each piece much match 👉 👈